frontend-react/src/utils/diff.ts

```diff
@@ -0,0 +1,183 @@
+/**
+ * The algorithm implemented here is based on "An O(NP) Sequence Comparison Algorithm"
+ * by described by Sun Wu, Udi Manber and Gene Myers
+ * Adapted from: https://github.com/cubicdaiya/onp
+ */
+
+export enum SES_TYPE {
+    DELETE = -1,
+    COMMON = 0,
+    ADD = 1,
+}
+
+export const Diff = (a_: string, b_: string) => {
+    type PEntryType = {
+        x: number;
+        y: number;
+        k: number;
+    };
+
+    type SesEntryType = {
+        sestype: SES_TYPE;
+        index: number;
+        len: number;
+    };
+
+    let a = a_,
+        b = b_,
+        alen = a.length,
+        blen = b.length,
+        reverse: boolean = false,
+        editdistance: number | null = null,
+        offset = alen + 1,
+        path: number[] = [],
+        pathposi: PEntryType[] = [],
+        ses: SesEntryType[] = [],
+        sesstate = SES_TYPE.COMMON;
+    let lcs = "";
+
+    const init = () => {
+        if (alen >= blen) {
+            // swap around a=b and b=a if a is longer
+            [a, alen, b, blen] = [b, blen, a, alen];
+
+            reverse = true;
+            offset = alen + 1;
+        }
+    };
+
+    const P = (x: number, y: number, k: number): PEntryType => {
+        return { x, y, k };
+    };
+
+    const addSesElem = (char: string, index: number, sestype: SES_TYPE) => {
+        const addnew =
+            ses.length === 0 || // first elem special case
+            sesstate !== sestype; // we switched, probably back to common.
+
+        sesstate = sestype; // save state to check for next character
+
+        if (addnew) {
+            // don't track common
+            ses.push({
+                sestype,
+                index,
+                len: 1,
+            });
+        } else {
+            ses[ses.length - 1].len++;
+        }
+    };
+
+    const snake = (k: number, p: number, pp: number): number => {
```

Comment:
🐍

Reply:
Heh... yeah, from the paper:

<img width="595" alt="image" src="https://user-images.githubusercontent.com/74203452/158904825-6c62d101-4cfb-4cf2-a222-4d31e9d0c1dc.png">